### DISCUSSION TITLE
Ability to modify device config files on the fly?

### DISCUSSION THREAD

**Discussion by NYZack** [zwave-js/zwave-js-ui#582] [Ideas]
Given that zwave js is in active development, it would be great to be able to modify or create device configuration files to fix gaps for particular devices - for instance, I have some motion sensors for which parameter settings haven't yet been implemented.  It seems like it would be fairly straightforward for me to implement this by referencing the Open Z-wave configuration files, without having to wait for the implementation to happen in the main zwave js code.

**Comment by robertsLando** [comment]
If you are using docker you can use the Docekfile.contrib file to build your own image with custom branches/edits, otherwise you can use the 'nodejs' way and clone/link the repositories locally

**Comment by ivdimitro** [comment]
> Given that zwave js is in active development, it would be great to be able to modify or create device configuration files to fix gaps for particular devices - for instance, I have some motion sensors for which parameter settings haven't yet been implemented. It seems like it would be fairly straightforward for me to implement this by referencing the Open Z-wave configuration files, without having to wait for the implementation to happen in the main zwave js code.

If you are using docker you can map the device folders from inside the container to local folder in which you can put your test files. For example: 
`     ./0x0371:/usr/src/app/node_modules/@zwave-js/config/config/devices/0x0371
      ./0x0086:/usr/src/app/node_modules/@zwave-js/config/config/devices/0x0086
      ./0x019b:/usr/src/app/node_modules/@zwave-js/config/config/devices/0x019b
      ./index.json:/usr/src/app/node_modules/@zwave-js/config/config/devices/index.json`

you will need to add index.json if you are adding new manufacturers.

**Reply by NYZack** [reply]
@robertsLando and @ivdimitro Yes, I've done this, and it works well.  My thought, though, is that Z-wave devices are constantly coming out and new device configuration files are constantly being put out there.  Expecting users to go through this (mildly complicated) process doesn't seem realistic.  This is especially the case as zwavejs2mqtt is being incorporated into an add-on for Home Assistant.  There's no straightforward way (as far as I can tell) to do this process for a Home Assistant add-on.

I think, for instance, about when I incorporated my Zooz ZSE29 motion sensor and my Inovelli LZW-31 dimmers into my system with the old Home Assistant Z-wave integration; these devices were not esoteric, but they were not (yet) incorporated into mainstream repositories.  I was able to download configuration files from the Internet and add them.  It took some know-how, but nothing like what it takes to separately start an independent docker container and bind-mount directories into it.

**Reply by robertsLando** [reply]
ATM running a `npm install @zwave-js/config` should do the trick on the fly. Correct @AlCalzone ?

**Reply by AlCalzone** [reply]
Yep, just update `@zwave-js/config` to get the latest release (as long as there were no breaking changes).

**Reply by NYZack** [reply]
Sure, but again hard to do in a Home Assistant add-on, and this doesn't account for configuration files that haven't been included yet in zwave-js.

**Reply by AlCalzone** [reply]
Since there is an open feature request for this: what **would** be an approach that also works for HA?

**Reply by NYZack** [reply]
Well, that's a harder question!  How about a way to copy-and-paste or upload a config file, which would then be placed in the appropriate config folder? There might be some rudimentary error checking (JSON sanity check, for instance).  

By the way, I noticed that when I added a custom config file that had a JSON syntax error and restarted zwavejs2mqtt, ALL of my devices showed up as "Unknown" (not just the one that had the bad config file).

**Reply by NYZack** [reply]
Another (simpler) option might be to place the device config folder in the exposed shared folder (the "store" folder), where the settings and cache files are currently stored.

**Reply by AlCalzone** [reply]
That was part of the plan - even better that it also works in the HA ecosystem.

> I noticed that when I added a custom config file that had a JSON syntax error and restarted zwavejs2mqtt, ALL of my devices showed up as "Unknown" (not just the one that had the bad config file).

Definitely not intended :( ==> https://github.com/zwave-js/node-zwave-js/issues/1825

**Reply by popescunsergiu** [reply]
zigbee2mqtt has a way in HA to make this possible by defining a list of device config files under property: external_converters: []

Can this be a way for Z-WaveJS to accept device config files so it's easier to test them before committing?

**Reply by AlCalzone** [reply]
@popescunsergiu did you see this?
https://zwave-js.github.io/zwavejs2mqtt/#/development/intro?id=testing-custom-devices-config

**Comment by AlCalzone** [comment]
This has been possible for a while now:
https://zwave-js.github.io/zwavejs2mqtt/#/development/intro?id=testing-custom-devices-config
